packages/lit-html/src/disconnectable-directive.ts

```diff
@@ -0,0 +1,382 @@
+ * it would still contain its directive D13, it stays in the disconnectable
+ * tree.
+ *
+ * Example 2: In the course of Example 1, `setConnected` will reach
+ * NodePart(N16); in this case the entire part is being disconnected, so we
+ * simply iterate all of N16's `_$disconnectableChildren` (D17,T18) and
+ * recursively run `setConnected` on them. Note that we only remove children
+ * from `_$disconnectableChildren` for the top-level values being disconnected
+ * on a clear; doing this bookkeeping lower in the tree is wasteful since it's
+ * all being thrown away.
+ *
+ * Example 3: If the LitElement containing the entire tree above becomes
+ * disconnected, it will run `nodePart.setConnected()` (which calls
+ * `nodePart._$setNodePartConnected()` if it exists); in this case, we
+ * recursively run `setConnected()` over the entire tree, without removing any
+ * children from `_$disconnectableChildren`, since this tree is required to
+ * re-connect the tree, which does the same operation, simply passing
+ * `isConnectd: true` down the tree, signaling which callback to run.
+ */
+
+import {
+  Directive,
+  PartInfo,
+  NodePart,
+  NODE_PART,
+  Disconnectable,
+  noChange,
+} from './lit-html.js';
+import {setPartValue} from './parts.js';
+export {directive} from './lit-html.js';
+
+/**
+ * Recursively walks down the tree of Parts/TemplateInstances/Directives to set
+ * the connected state of directives and run `disconnectedCallback`/
+ * `reconnectedCallback`s.
+ *
+ * @returns True if there were children to disconnect; false otherwise
+ */
+const setChildrenConnected = (
+  parent: Disconnectable,
+  isConnected: boolean
+): boolean => {
+  const children = parent._$disconnetableChildren;
+  if (children === undefined) {
+    return false;
+  }
+  for (const obj of children) {
+    // The existence of `_$setDirectiveConnected` is used as a "brand" to
+    // disambiguate DisconnectableDirectives from other DisconnectableChildren
+    // (as opposed to using an instanceof check to know when to call it); the
+    // redundancy of "Directive" in the API name is to avoid conflicting with
+    // `_$setNodePartConnected`, which exists `NodeParts` which are also in
+    // this list
+    // Disconnect Directive (and any nested directives contained within)
+    (obj as DisconnectableDirective)._$setDirectiveConnected?.(
+      isConnected,
+      false
+    );
+    // Disconnect Part/TemplateInstance
+    setChildrenConnected(obj, isConnected);
+  }
+  return true;
+};
+
+/**
+ * Removes the given child from its parent list of disconnectable children, and
+ * if the parent list becomes empty as a result, removes the parent from its
+ * parent, and so forth up the tree when that causes subsequent parent lists to
+ * become empty.
+ */
+const removeDisconnectableFromParent = (obj: Disconnectable) => {
+  let parent, children;
+  do {
+    if ((parent = obj._$parent) == undefined) {
```

Comment:
use `===` here, since this can't be `null`

Reply:
Done